File: go/236.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	if root == nil || root == p || root == q {
		return root
	}
	var l, r *TreeNode = lowestCommonAncestor(root.Left, p, q), lowestCommonAncestor(root.Right, p, q)
	if l != nil && r != nil {
	  return root
	} else if l != nil {
	  return l
	} else if r != nil {
	  return r
	}
	return nil
}

File: cpp/236.cpp
#include <iostream>

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  TreeNode *find(TreeNode *r, TreeNode *p, TreeNode *q) {
    if (!r || r == p || r == q)
      return r;
    TreeNode *left = find(r->left, p, q), *right = find(r->right, p, q);
    if (left && right)
      return r;
    else if (left)
      return left;
    else if (right)
      return right;
    return nullptr;
  }
  TreeNode *lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q) {
    return find(root, p, q);
  }
};
